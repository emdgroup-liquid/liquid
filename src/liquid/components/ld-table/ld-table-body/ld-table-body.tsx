import { Component, h } from '@stencil/core'

/**
 * @part tbody - the table body
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-table-body',
  styleUrl: 'ld-table-body.shadow.css',
  shadow: true,
})
export class LdTableBody {
  render() {
    return (
      <tbody class="ld-table-body" part="tbody">
        <slot />
      </tbody>
    )
  }
}
