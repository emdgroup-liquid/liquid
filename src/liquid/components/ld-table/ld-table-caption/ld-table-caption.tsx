import { Component, h } from '@stencil/core'

/**
 * @part figcaption - the actual figcaption element
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-table-caption',
  styleUrl: 'ld-table-caption.shadow.css',
  shadow: true,
})
export class LdTableCaption {
  render() {
    return (
      <figcaption class="ld-table-caption" part="figcaption">
        <slot></slot>
      </figcaption>
    )
  }
}
