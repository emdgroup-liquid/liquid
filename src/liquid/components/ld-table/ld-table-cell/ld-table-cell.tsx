import { Component, h, Prop } from '@stencil/core'

/**
 * @part cell - the actual td element
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-table-cell',
  styleUrl: 'ld-table-cell.shadow.css',
  shadow: true,
})
export class LdTableCell {
  /** Contains a non-negative integer value that indicates for how many columns the cell extends. */
  @Prop() colspan: HTMLTableCellElement['colSpan']

  /** Contains a list of space-separated strings, each corresponding to the id attribute of the table header elements that apply to this element. */
  @Prop() headers: HTMLTableCellElement['headers']

  /** Contains a non-negative integer value that indicates for how many rows the cell extends. */
  @Prop() rowspan: HTMLTableCellElement['rowSpan']

  render() {
    return (
      <td
        class="ld-table-cell"
        colSpan={this.colspan}
        headers={this.headers}
        part="cell"
        rowSpan={this.rowspan}
      >
        <slot />
      </td>
    )
  }
}
