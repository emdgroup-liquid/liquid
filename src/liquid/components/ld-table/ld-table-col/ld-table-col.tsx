import { Component, h, Prop } from '@stencil/core'

/**
 * @part col - the col
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-table-col',
  styleUrl: 'ld-table-col.shadow.css',
  shadow: true,
})
export class LdTableCol {
  /** indicating the number of consecutive columns the colgroup element spans. */
  @Prop() span: HTMLTableColElement['span']

  render() {
    return (
      <col class="ld-table-col" span={this.span} part="col">
        <slot />
      </col>
    )
  }
}
