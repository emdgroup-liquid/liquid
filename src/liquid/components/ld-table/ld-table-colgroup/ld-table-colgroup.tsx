import { Component, h, Prop } from "@stencil/core";

/**
 * @part colgroup - the actual colgroup element
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: "ld-table-colgroup",
  styleUrl: "ld-table-colgroup.shadow.css",
  shadow: true,
})
export class LdTableColgroup {
  /** Contains a non-negative integer value indicating the number of consecutive columns the colgroup element spans. */
  @Prop() span?: HTMLTableColElement["span"];

  render() {
    return (
      <colgroup class="ld-table-colgroup" span={this.span} part="colgroup">
        <slot />
      </colgroup>
    );
  }
}
