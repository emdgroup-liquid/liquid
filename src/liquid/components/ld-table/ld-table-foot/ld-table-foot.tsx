import { Component, h } from '@stencil/core'

/**
 * @part tfoot - the table footer
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-table-foot',
  styleUrl: 'ld-table-foot.shadow.css',
  shadow: true,
})
export class LdTableFoot {
  render() {
    return (
      <tfoot class="ld-table-foot" part="tfoot">
        <slot />
      </tfoot>
    )
  }
}
