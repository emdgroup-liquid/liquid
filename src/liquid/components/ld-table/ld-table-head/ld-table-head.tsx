import { Component, h } from "@stencil/core";

/**
 * @part thead - the actual thead element
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: "ld-table-head",
  styleUrl: "ld-table-head.shadow.css",
  shadow: true,
})
export class LdTableHead {
  render() {
    return (
      <thead class="ld-table-head" part="thead">
        <slot />
      </thead>
    );
  }
}
