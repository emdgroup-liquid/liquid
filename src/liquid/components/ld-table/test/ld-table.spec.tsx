import '../../../utils/mutationObserver'
import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdCheckbox } from '../../ld-checkbox/ld-checkbox'
import { LdTable } from '../ld-table'
import { LdTableBody } from '../ld-table-body/ld-table-body'
import { LdTableCaption } from '../ld-table-caption/ld-table-caption'
import { LdTableCell } from '../ld-table-cell/ld-table-cell'
import { LdTableCol } from '../ld-table-col/ld-table-col'
import { LdTableColgroup } from '../ld-table-colgroup/ld-table-colgroup'
import { LdTableFoot } from '../ld-table-foot/ld-table-foot'
import { LdTableHeader } from '../ld-table-header/ld-table-header'
import { LdTableHead } from '../ld-table-head/ld-table-head'
import { LdTableRow } from '../ld-table-row/ld-table-row'
import { LdTableToolbar } from '../ld-table-toolbar/ld-table-toolbar'

const components = [
  LdCheckbox,
  LdTable,
  LdTableBody,
  LdTableCaption,
  LdTableCell,
  LdTableCol,
  LdTableColgroup,
  LdTableFoot,
  LdTableHead,
  LdTableHeader,
  LdTableRow,
  LdTableToolbar,
]

describe('ld-table', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components,
      template: () => (
        <ld-table>
          <ld-table-toolbar slot="toolbar">
            <ld-table-caption>Superheros and sidekicks</ld-table-caption>
          </ld-table-toolbar>
          <ld-table-colgroup>
            <ld-table-col></ld-table-col>
            <ld-table-col span={2} class="batman"></ld-table-col>
            <ld-table-col span={2} class="flash"></ld-table-col>
          </ld-table-colgroup>
          <ld-table-body>
            <ld-table-row>
              <ld-table-cell></ld-table-cell>
              <ld-table-header scope="col">Batman</ld-table-header>
              <ld-table-header scope="col">Robin</ld-table-header>
              <ld-table-header scope="col">The Flash</ld-table-header>
              <ld-table-header scope="col">Kid Flash</ld-table-header>
            </ld-table-row>
            <ld-table-row>
              <ld-table-header scope="row">Skill</ld-table-header>
              <ld-table-cell>Smarts</ld-table-cell>
              <ld-table-cell>Dex, acrobat</ld-table-cell>
              <ld-table-cell>Super speed</ld-table-cell>
              <ld-table-cell>Super speed</ld-table-cell>
            </ld-table-row>
          </ld-table-body>
        </ld-table>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  describe('sorting', () => {
    it('sorts', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row>
                <ld-table-header sortable sort-order="asc">
                  Nums
                </ld-table-header>
                <ld-table-header sortable>Chars</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row>
                <ld-table-cell>1.0</ld-table-cell>
                <ld-table-cell>A</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>200,000.0</ld-table-cell>
                <ld-table-cell>B</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>100,000,000.0</ld-table-cell>
                <ld-table-cell>C</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      const getNums = () =>
        Array.from(page.root.querySelectorAll('ld-table-cell:first-child')).map(
          (cell) => cell.textContent
        )
      let nums = getNums()
      const getChars = () =>
        Array.from(
          page.root.querySelectorAll('ld-table-cell:nth-child(2)')
        ).map((cell) => cell.textContent)
      let chars = getChars()

      expect(nums[0]).toEqual('1.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('100,000,000.0')

      const btnNumsSortAsc = page.root
        .querySelector('ld-table-header:first-child')
        .shadowRoot.querySelector(
          '[part="sort-button-asc"]'
        ) as HTMLLdButtonElement
      const btnNumsSortDesc = page.root
        .querySelector('ld-table-header:first-child')
        .shadowRoot.querySelector(
          '[part="sort-button-desc"]'
        ) as HTMLLdButtonElement

      const btnCharsSortAsc = page.root
        .querySelector('ld-table-header:nth-child(2)')
        .shadowRoot.querySelector(
          '[part="sort-button-asc"]'
        ) as HTMLLdButtonElement
      const btnCharsSortDesc = page.root
        .querySelector('ld-table-header:nth-child(2)')
        .shadowRoot.querySelector(
          '[part="sort-button-desc"]'
        ) as HTMLLdButtonElement

      expect(btnNumsSortAsc).toHaveAttribute('aria-disabled')
      btnNumsSortAsc.ariaDisabled = 'true'
      expect(btnNumsSortDesc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')

      btnNumsSortAsc.click()
      await page.waitForChanges()
      nums = getNums()
      chars = getChars()

      expect(nums[0]).toEqual('1.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('100,000,000.0')
      expect(chars[0]).toEqual('A')
      expect(chars[1]).toEqual('B')
      expect(chars[2]).toEqual('C')

      expect(btnNumsSortAsc).toHaveAttribute('aria-disabled')
      expect(btnNumsSortDesc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')

      btnNumsSortDesc.click()
      await page.waitForChanges()
      nums = getNums()
      chars = getChars()

      expect(nums[0]).toEqual('100,000,000.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('1.0')
      expect(chars[0]).toEqual('C')
      expect(chars[1]).toEqual('B')
      expect(chars[2]).toEqual('A')

      expect(btnNumsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnNumsSortDesc).toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')

      btnCharsSortAsc.click()
      await page.waitForChanges()
      nums = getNums()
      chars = getChars()

      expect(nums[0]).toEqual('1.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('100,000,000.0')
      expect(chars[0]).toEqual('A')
      expect(chars[1]).toEqual('B')
      expect(chars[2]).toEqual('C')

      expect(btnNumsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnNumsSortDesc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')

      const th = page.root
        .querySelector('ld-table-header:first-child')
        .shadowRoot.querySelector('th')

      th.click()
      await page.waitForChanges()
      nums = getNums()
      chars = getChars()

      expect(nums[0]).toEqual('100,000,000.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('1.0')
      expect(chars[0]).toEqual('C')
      expect(chars[1]).toEqual('B')
      expect(chars[2]).toEqual('A')

      expect(btnNumsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnNumsSortDesc).toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')

      th.click()
      await page.waitForChanges()
      nums = getNums()
      chars = getChars()

      expect(nums[0]).toEqual('1.0')
      expect(nums[1]).toEqual('200,000.0')
      expect(nums[2]).toEqual('100,000,000.0')
      expect(chars[0]).toEqual('A')
      expect(chars[1]).toEqual('B')
      expect(chars[2]).toEqual('C')

      expect(btnNumsSortAsc).toHaveAttribute('aria-disabled')
      expect(btnNumsSortDesc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortAsc).not.toHaveAttribute('aria-disabled')
      expect(btnCharsSortDesc).not.toHaveAttribute('aria-disabled')
    })

    it('does not sort if prevented', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row>
                <ld-table-header sortable sort-order="asc">
                  Nums
                </ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
            <ld-table-foot>
              <ld-table-row>
                <ld-table-cell>6</ld-table-cell>
              </ld-table-row>
            </ld-table-foot>
          </ld-table>
        ),
      })

      const getNums = () =>
        Array.from(page.root.querySelectorAll('ld-table-cell:first-child')).map(
          (cell) => cell.textContent
        )
      let nums = getNums()

      expect(nums[0]).toEqual('1')
      expect(nums[1]).toEqual('2')
      expect(nums[2]).toEqual('3')

      page.root.querySelector('ld-table-header').dispatchEvent(
        new CustomEvent('ldTableSort', {
          bubbles: true,
          detail: {
            columnIndex: 0,
            sortOrder: 'desc',
          },
        })
      )
      await page.waitForChanges()
      nums = getNums()

      expect(nums[0]).toEqual('3')
      expect(nums[1]).toEqual('2')
      expect(nums[2]).toEqual('1')

      const prevented = new CustomEvent('ldTableSort', {
        bubbles: true,
        detail: {
          columnIndex: 0,
          sortOrder: 'asc',
        },
      })
      prevented.preventDefault()

      page.root.querySelector('ld-table-header').dispatchEvent(prevented)
      await page.waitForChanges()
      nums = getNums()

      expect(nums[0]).toEqual('3')
      expect(nums[1]).toEqual('2')
      expect(nums[2]).toEqual('1')
    })

    it('does not sort if not sortable', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row>
                <ld-table-header sort-order="asc">Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
            <ld-table-foot>
              <ld-table-row>
                <ld-table-cell>6</ld-table-cell>
              </ld-table-row>
            </ld-table-foot>
          </ld-table>
        ),
      })

      const getNums = () =>
        Array.from(page.root.querySelectorAll('ld-table-cell:first-child')).map(
          (cell) => cell.textContent
        )
      let nums = getNums()

      expect(nums[0]).toEqual('1')
      expect(nums[1]).toEqual('2')
      expect(nums[2]).toEqual('3')

      const th = page.root
        .querySelector('ld-table-header:first-child')
        .shadowRoot.querySelector('th')

      th.click()
      await page.waitForChanges()
      nums = getNums()

      expect(nums[0]).toEqual('1')
      expect(nums[1]).toEqual('2')
      expect(nums[2]).toEqual('3')
    })
  })

  describe('selection', () => {
    it('selects', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row selectable>
                <ld-table-header>Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      const getCheckboxes = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) =>
          row.shadowRoot.querySelector('ld-checkbox')
        )
      const checkboxes = getCheckboxes()
      expect(checkboxes.length).toEqual(4)

      expect(checkboxes[0].indeterminate).toBeFalsy()
      expect(checkboxes[0].checked).toBeFalsy()
      expect(checkboxes[1].checked).toBeFalsy()
      expect(checkboxes[2].checked).toBeFalsy()
      expect(checkboxes[3].checked).toBeFalsy()

      checkboxes[2].click()
      await page.waitForChanges()

      expect(checkboxes[0].indeterminate).toBeTruthy()
      expect(checkboxes[0].checked).toBeFalsy()
      expect(checkboxes[1].checked).toBeFalsy()
      expect(checkboxes[2].checked).toBeTruthy()
      expect(checkboxes[3].checked).toBeFalsy()

      checkboxes[0].click()
      await page.waitForChanges()

      expect(checkboxes[0].indeterminate).toBeFalsy()
      expect(checkboxes[0].checked).toBeTruthy()
      expect(checkboxes[1].checked).toBeTruthy()
      expect(checkboxes[2].checked).toBeTruthy()
      expect(checkboxes[3].checked).toBeTruthy()

      checkboxes[2].click()
      await page.waitForChanges()

      expect(checkboxes[0].indeterminate).toBeTruthy()
      expect(checkboxes[0].checked).toBeFalsy()
      expect(checkboxes[1].checked).toBeTruthy()
      expect(checkboxes[2].checked).toBeFalsy()
      expect(checkboxes[3].checked).toBeTruthy()

      checkboxes[2].click()
      await page.waitForChanges()

      expect(checkboxes[0].indeterminate).toBeFalsy()
      expect(checkboxes[0].checked).toBeTruthy()
      expect(checkboxes[1].checked).toBeTruthy()
      expect(checkboxes[2].checked).toBeTruthy()
      expect(checkboxes[3].checked).toBeTruthy()
    })

    it('does not select if prevented', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row selectable>
                <ld-table-header>Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      const getRows = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) => ({
          row,
          checkbox: row.shadowRoot.querySelector('ld-checkbox'),
        }))
      const rows = getRows()

      rows[2].row.selected = true
      rows[2].checkbox.checked = true
      page.root.dispatchEvent(
        new CustomEvent('ldTableSelect', {
          bubbles: true,
          detail: {
            rowIndex: 2,
            selected: true,
          },
        })
      )
      await page.waitForChanges()

      expect(rows[0].checkbox.indeterminate).toBeTruthy()

      rows[2].row.selected = false
      rows[2].checkbox.checked = false
      const prevented = new CustomEvent('ldTableSelect', {
        bubbles: true,
        detail: {
          rowIndex: 2,
          selected: false,
        },
      })
      prevented.preventDefault()
      page.root.dispatchEvent(prevented)

      expect(rows[0].checkbox.indeterminate).toBeTruthy()
    })

    it('does not select all if prevented', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row selectable>
                <ld-table-header>Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      const getCheckboxes = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) =>
          row.shadowRoot.querySelector('ld-checkbox')
        )
      const checkboxes = getCheckboxes()

      page.root.dispatchEvent(
        new CustomEvent('ldTableSelectAll', {
          bubbles: true,
          detail: {
            selected: true,
          },
        })
      )
      await page.waitForChanges()

      expect(checkboxes[1].checked).toBeTruthy()
      expect(checkboxes[2].checked).toBeTruthy()
      expect(checkboxes[3].checked).toBeTruthy()

      page.root.dispatchEvent(
        new CustomEvent('ldTableSelectAll', {
          bubbles: true,
          detail: {
            selected: false,
          },
        })
      )
      await page.waitForChanges()

      expect(checkboxes[1].checked).toBeFalsy()
      expect(checkboxes[2].checked).toBeFalsy()
      expect(checkboxes[3].checked).toBeFalsy()

      const prevented = new CustomEvent('ldTableSelectAll', {
        bubbles: true,
        detail: {
          selected: true,
        },
      })
      prevented.preventDefault()
      page.root.dispatchEvent(prevented)

      expect(checkboxes[1].checked).toBeFalsy()
      expect(checkboxes[2].checked).toBeFalsy()
      expect(checkboxes[3].checked).toBeFalsy()
    })

    it('updates select all checkbox on load', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row selectable>
                <ld-table-header>Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable selected>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      await page.waitForChanges()
      const getCheckboxes = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) =>
          row.shadowRoot.querySelector('ld-checkbox')
        )
      const checkboxes = getCheckboxes()

      expect(checkboxes[0].indeterminate).toBeTruthy()
    })

    it('does not throw updating select all checkbox on load if there is no table head', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable selected>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })

      await page.waitForChanges()
      const getCheckboxes = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) =>
          row.shadowRoot.querySelector('ld-checkbox')
        )
      const checkboxes = getCheckboxes()

      expect(checkboxes.length).toEqual(3)
    })

    it('disabled select all if at least one row has a disabled selection', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-table>
            <ld-table-head>
              <ld-table-row selectable>
                <ld-table-header>Nums</ld-table-header>
              </ld-table-row>
            </ld-table-head>
            <ld-table-body>
              <ld-table-row selectable>
                <ld-table-cell>1</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable selectionDisabled>
                <ld-table-cell>2</ld-table-cell>
              </ld-table-row>
              <ld-table-row selectable>
                <ld-table-cell>3</ld-table-cell>
              </ld-table-row>
            </ld-table-body>
          </ld-table>
        ),
      })
      await page.waitForChanges()

      const getCheckboxes = () =>
        Array.from(page.root.querySelectorAll('ld-table-row')).map((row) =>
          row.shadowRoot.querySelector('ld-checkbox')
        )
      const checkboxes = getCheckboxes()

      expect(checkboxes[0]).toHaveAttribute('aria-disabled')
    })
  })
})
