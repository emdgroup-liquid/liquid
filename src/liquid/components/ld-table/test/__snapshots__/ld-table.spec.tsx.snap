// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-table renders 1`] = `
<ld-table class="ld-table" role="figure">
  <mock:shadow-root>
    <slot name="toolbar"></slot>
    <div class="ld-table__scroll-container" part="scroll-container">
      <table part="table">
        <slot></slot>
      </table>
    </div>
  </mock:shadow-root>
  <ld-table-toolbar class="ld-table__toolbar" slot="toolbar">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <ld-table-caption>
      <mock:shadow-root>
        <figcaption class="ld-table-caption" part="figcaption">
          <slot></slot>
        </figcaption>
      </mock:shadow-root>
      Superheros and sidekicks
    </ld-table-caption>
  </ld-table-toolbar>
  <ld-table-colgroup>
    <mock:shadow-root>
      <colgroup class="ld-table-colgroup" part="colgroup">
        <slot></slot>
      </colgroup>
    </mock:shadow-root>
    <ld-table-col>
      <mock:shadow-root>
        <col class="ld-table-col" part="col">
      </mock:shadow-root>
    </ld-table-col>
    <ld-table-col class="batman">
      <mock:shadow-root>
        <col class="ld-table-col" part="col" span="2">
      </mock:shadow-root>
    </ld-table-col>
    <ld-table-col class="flash">
      <mock:shadow-root>
        <col class="ld-table-col" part="col" span="2">
      </mock:shadow-root>
    </ld-table-col>
  </ld-table-colgroup>
  <ld-table-body>
    <mock:shadow-root>
      <tbody class="ld-table-body" part="tbody">
        <slot></slot>
      </tbody>
    </mock:shadow-root>
    <ld-table-row>
      <mock:shadow-root>
        <tr class="ld-table-row" part="row">
          <slot></slot>
        </tr>
      </mock:shadow-root>
      <ld-table-cell>
        <mock:shadow-root>
          <td class="ld-table-cell" part="cell">
            <slot></slot>
          </td>
        </mock:shadow-root>
      </ld-table-cell>
      <ld-table-header>
        <mock:shadow-root>
          <th class="ld-table-header" part="cell" scope="col">
            <slot></slot>
          </th>
        </mock:shadow-root>
        Batman
      </ld-table-header>
      <ld-table-header>
        <mock:shadow-root>
          <th class="ld-table-header" part="cell" scope="col">
            <slot></slot>
          </th>
        </mock:shadow-root>
        Robin
      </ld-table-header>
      <ld-table-header>
        <mock:shadow-root>
          <th class="ld-table-header" part="cell" scope="col">
            <slot></slot>
          </th>
        </mock:shadow-root>
        The Flash
      </ld-table-header>
      <ld-table-header>
        <mock:shadow-root>
          <th class="ld-table-header" part="cell" scope="col">
            <slot></slot>
          </th>
        </mock:shadow-root>
        Kid Flash
      </ld-table-header>
    </ld-table-row>
    <ld-table-row>
      <mock:shadow-root>
        <tr class="ld-table-row" part="row">
          <slot></slot>
        </tr>
      </mock:shadow-root>
      <ld-table-header>
        <mock:shadow-root>
          <th class="ld-table-header" part="cell" scope="row">
            <slot></slot>
          </th>
        </mock:shadow-root>
        Skill
      </ld-table-header>
      <ld-table-cell>
        <mock:shadow-root>
          <td class="ld-table-cell" part="cell">
            <slot></slot>
          </td>
        </mock:shadow-root>
        Smarts
      </ld-table-cell>
      <ld-table-cell>
        <mock:shadow-root>
          <td class="ld-table-cell" part="cell">
            <slot></slot>
          </td>
        </mock:shadow-root>
        Dex, acrobat
      </ld-table-cell>
      <ld-table-cell>
        <mock:shadow-root>
          <td class="ld-table-cell" part="cell">
            <slot></slot>
          </td>
        </mock:shadow-root>
        Super speed
      </ld-table-cell>
      <ld-table-cell>
        <mock:shadow-root>
          <td class="ld-table-cell" part="cell">
            <slot></slot>
          </td>
        </mock:shadow-root>
        Super speed
      </ld-table-cell>
    </ld-table-row>
  </ld-table-body>
</ld-table>
`;
